ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Restore.hs

```diff
@@ -0,0 +1,128 @@
+-- | Cardano Wallet restoration logic.
+
+module Ariadne.Wallet.Cardano.Kernel.Restore
+       ( getKeyFromMnemonic
+       , getUtxoByAccount
+       , readNonAriadneKeys
+       ) where
+
+import qualified Universum.Unsafe as Unsafe (init)
+
+import Control.Exception (Exception(displayException))
+import Control.Lens (at, non, (?~))
+import Control.Natural (type (~>))
+import qualified Data.ByteString as BS
+import qualified Data.Map.Strict as Map
+
+import Pos.Binary.Class (decodeFull')
+import Pos.Core.Configuration (HasConfiguration)
+import Pos.Crypto (EncryptedSecretKey, PassPhrase)
+import qualified Pos.Crypto as Crypto
+  (deriveHDPassphrase, encToPublic, safeDeterministicKeyGen)
+import Pos.Txp.Toil.Types (Utxo)
+import Pos.Util.BackupPhrase (BackupPhrase(..), safeKeysFromPhrase)
+import Pos.Util.UserSecret (usKeys0)
+
+import Ariadne.Cardano.Face (Address, CardanoMode)
+import Ariadne.Wallet.Backend.AddressDiscovery
+  (AddressWithPathToUtxoMap, discoverHDAddressWithUtxo)
+import Ariadne.Wallet.Cardano.Kernel.Bip32 (DerivationPath(..))
+import Ariadne.Wallet.Cardano.Kernel.Bip39 (mnemonicToSeedNoPassword)
+import Ariadne.Wallet.Cardano.Kernel.Bip44
+  (Bip44DerivationPath(..), bip44PathToAddressId, decodeBip44DerivationPath)
+import Ariadne.Wallet.Cardano.Kernel.DB.HdWallet
+  (HdAddressId, eskToHdRootId, hdAddressIdParent)
+import Ariadne.Wallet.Cardano.Kernel.PrefilterTx
+  (PrefilteredUtxo, UtxoByAccount)
+import Ariadne.Wallet.Face (WalletRestoreType(..))
+
+newtype WrongMnemonic = WrongMnemonic Text
+ deriving (Eq, Show)
+
+instance Exception WrongMnemonic where
+  displayException (WrongMnemonic txt) =
+    "Wrong mnemonic: " <> show txt
+
+data SecretsDecodingError = SecretsDecodingError FilePath Text
```

Comment:
These exception types were copied from the other `Restore.hs` module, they are not used there anymore, so they had to be moved instead, not copied.